### PR TITLE
feat: Pagination on UserPage

### PR DESCRIPTION
**Pull request by Nimasheth**
Add Pagination To Table Component

[Add pagination to Table component](https://www.notion.so/uwblueprintexecs/Add-pagination-to-Table-component-9d2ce24f58814d0da04eefc013abca9b)

Adds pagination component to table component!

## Checklist
- [X] My PR name is descriptive and in imperative tense
- [X] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [X] I have run the appropriate linter(s)
- [X] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -27,12 +28,14 @@
     "axios": "^0.21.1",
     "bootstrap": "^4.6.0",
     "chakra-react-select": "^4.4.2",
+    "classnames": "^2.3.2",
```
Comment (carissa-tang) [inline frontend/package.json]: can you remove the unused libraries? (i.e. `yarn remove classnames prop-types`)

File: frontend/src/themes/components/button.ts [hunk 2]
```diff
@@ -51,7 +51,7 @@ const Button = {
       WhiteIcon,
     },
     secondary: {
-      bg: "blue.50",
+      bg: "#154472",
```
Comment (carissa-tang) [inline frontend/src/themes/components/button.ts]: is this change necessary?

File: frontend/src/components/common/Pagination.tsx [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    heading: '"DM Sans"',
+    mono: '"DM Sans"',
+  },
+  colors: {
+    brand: {
+      900: "#154472",
+    },
+  },
+  components: {
+    PaginationSeparator: {
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]: we already have a custom theme! can you move this into `themes/components/paginationSeparator.ts`

Comment (carissa-tang) [reply frontend/src/components/common/Pagination.tsx]: you will also need to update `themes.index.ts`

**Review by carissa-tang — COMMENTED** [review]
looks good so far! left some comments for refactoring

also, could you pull from staging ? so your changes are not mixed with old changes

**Comment by Nimasheth** [comment]
https://user-images.githubusercontent.com/60047036/219285391-01d10a53-b9ef-4272-8a21-f3ecbdad992b.MOV

can change the number of items per page, still have to ask design how many we are thinking! left it at 8 per page right now.

File: frontend/package.json [hunk 4]
```diff
@@ -33,6 +34,7 @@
     "json-schema": "^0.3.0",
     "json2csv": "^5.0.6",
     "jsonwebtoken": "^8.5.1",
+    "prop-types": "^15.8.1",
```
Comment (carissa-tang) [inline frontend/package.json]: yarn remove this

File: frontend/src/components/common/Pagination.tsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  PaginationNext,
+  PaginationPage,
+  PaginationPrevious,
+  PaginationContainer,
+  PaginationPageGroup,
+  PaginationSeparator,
+} from "@ajna/pagination";
+import pagination from "../../themes/components/pagination";
+
+const customTheme = {
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]: you can remove this and add `pagination` to `themes/index.ts`

File: frontend/src/themes/components/pagination.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+const pagination = {
```
Comment (carissa-tang) [inline frontend/src/themes/components/pagination.ts]: can you capitalize this? for convention

File: frontend/src/components/common/Table.tsx [hunk 7]
```diff
@@ -8,51 +8,72 @@ import {
...
 } from "@chakra-ui/react";
+import Pagination from "./Pagination";
 
 export interface TableRow {
   values: any[];
   menu: React.ReactElement;
 }
 
-interface TableProps {
+export interface TableProps {
```
Comment (carissa-tang) [inline frontend/src/components/common/Table.tsx]:
```suggestion
interface TableProps {
```
do you need this change?

File: frontend/src/components/common/Pagination.tsx (hunk 5 above)
Comment (Nimasheth) [reply frontend/src/components/common/Pagination.tsx]: when i do that, it messes up all the buttons and how they look so I have to import pagination into the Pagination.ts file. I tried to import the theme we already had with the pagination style added to it but that gave the same result. Unless there is another way that I don't of that allows me to add pagination to themes/index.ts while not messing up the styling of the pagination on the table page!?

File: frontend/src/themes/index.ts [hunk 8]
```diff
@@ -25,6 +26,7 @@ const theme = extendTheme({
     Input,
     Radio,
     Table,
+    paginationStyle,
```
Comment (carissa-tang) [inline frontend/src/themes/index.ts]: sorry for the nit, do you mind renaming this to `Pagination` (capitalized)?

File: frontend/src/components/common/Pagination.tsx [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+              {pages.map((page: number, index: number) => {
+                if (
+                  index === 0 ||
+                  index === pages.length - 1 ||
+                  (index >= currentPage - innerLimit - 1 &&
+                    index <= currentPage + innerLimit - 1)
+                ) {
+                  return (
+                    <PaginationPage
+                      bg="#E8EDF1"
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]: rather than hard coding the colors, we like to use their aliases (defined in `themes/colors.ts`) e.g. `"#E8EDF1" = "blue.50"`

could you update all instances where you hard coded the colors?

File: frontend/src/components/common/Pagination.tsx [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+                      key={`pagination_page_${page}`}
+                      page={page}
+                      onClick={() => console.warn("Im clicking the page")}
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]: this is great for testing! but once it works we can remove these debugging comments
```suggestion
```

File: frontend/src/components/common/Pagination.tsx (hunk 5 above)
Comment (carissa-tang) [reply frontend/src/components/common/Pagination.tsx]: I think you can remove this entirely and remove the `<ChakraProvider theme={customTheme}>` tag in the component (the theme is already applied at the top level). 

It will look a bit wonky though because our default style for buttons have `minWidth: "180px",`. To solve this, you can actually define a new button size and variant by moving all the styling code for `PaginationPage` under `themes/components/button.ts`
image

File: frontend/src/components/common/Pagination.tsx [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+                      color="#636363"
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]: as explained above, please move all this into `themes/components/button.ts` and create a new variant and size

then you can do smth like
```
<PaginationPage
   size="paginationSize"
   variant="pagination"
   ...
/>
```

File: frontend/src/themes/components/button.ts [hunk 12]
```diff
@@ -27,8 +29,70 @@ const Button = {
...
+      bg: "transparent",
+      color: "grey.300",
+    },
+
+    paginationPrevious: {
+      bg: "transparent",
+      color: "grey.300",
+    },
+
+    PaginationStyle: {
```
Comment (carissa-tang) [inline frontend/src/themes/components/button.ts]:
```suggestion
    paginationStyle: {
```

File: frontend/src/themes/components/button.ts [hunk 13]
```diff
@@ -1,3 +1,5 @@
+import { PaginationPrevious } from "@ajna/pagination";
```
Comment (carissa-tang) [inline frontend/src/themes/components/button.ts]:
```suggestion
```

File: frontend/src/themes/components/button.ts [hunk 14]
```diff
@@ -27,8 +29,70 @@ const Button = {
...
+      },
+    },
   },
+
   variants: {
+    paginationNext: {
+      bg: "transparent",
+      color: "grey.300",
+    },
+
```
Comment (carissa-tang) [inline frontend/src/themes/components/button.ts]: can you remove the extra spacing between brackets in this file?

File: frontend/src/themes/components/button.ts [hunk 15]
```diff
@@ -27,8 +29,70 @@ const Button = {
...
+      marginRight: "10px",
+      style: {
+        height: "75px",
+        margin: "0 10px",
+      },
+    },
   },
+
   variants: {
+    paginationNext: {
```
Comment (carissa-tang) [inline frontend/src/themes/components/button.ts]: since `paginationNext` and `paginationPrevious` have the same styling you can just use one and call it smth like `paginationNavigate` (for both the variant and size)

File: frontend/src/themes/components/button.ts [hunk 16]
```diff
@@ -27,8 +29,70 @@ const Button = {
...
+      fontWeight: "normal",
+      border: "none",
+      textDecoration: "underline",
+      outline: "none",
+      _hover: { textDecoration: "underline", fontWeight: "600" },
+      marginLeft: "10px",
+      style: {
+        height: "75px",
+        margin: "0 10px",
+      },
```
Comment (carissa-tang) [inline frontend/src/themes/components/button.ts]:
```suggestion
      fontWeight: "400",
      fontSize: "12px",
      lineHeight: "16px",
      textDecoration: "underline",
      _hover: { 
        fontWeight: "600" 
      },
      margin: "0 10px",
```

File: frontend/src/themes/components/button.ts [hunk 17]
```diff
@@ -27,8 +29,70 @@ const Button = {
...
+      _active: {
+        borderRadius: "50%",
+      },
+      height: "2.25rem",
+      width: "2.25rem",
+      borderRadius: "50%",
+      textAlign: "center",
+      lineHeight: "2rem",
+      display: "inline-block",
+      margin: "0 2px",
```
Comment (carissa-tang) [inline frontend/src/themes/components/button.ts]:
```suggestion
      height: "2rem",
      width: "2rem",
      borderRadius: "50%",
      margin: "0 2px",
      fontWeight: "700",
      fontSize: "12px",
```
generally, you want to only add styling if it's not there by default. also, make sure the styles are consistent with Figma

File: frontend/src/components/common/Table.tsx [hunk 18]
```diff
@@ -20,39 +21,59 @@ interface TableProps {
 }
 
 export const Table = ({ headers, rows }: TableProps): React.ReactElement => {
+  const [currentPage, setCurrentPage] = useState(1);
+  const itemsPerPage = 8; // can edit this to show how many itemsperpage we want
```
Comment (carissa-tang) [inline frontend/src/components/common/Table.tsx]: creds to @joyce-shi for this comment:

would you be able to remove the extraneous comments here and below? in general, it's best practice not to have comments duplicate / describe the code functionality. instead, the code should be clear enough to speak for itself!

you can read more on this here: https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/

File: frontend/src/components/common/Table.tsx [hunk 19]
```diff
@@ -20,39 +21,59 @@ interface TableProps {
...
-          ))}
-        </Tbody>
-      </T>
-    </TableContainer>
+          </Thead>
+          <Tbody>
+            {itemsToShow.map((row, rowIndex) => (
+              <Tr
+                key={rowIndex}
+                backgroundColor={rowIndex % 2 === 0 ? "#E8EDF1" : "#FFFFFF"}
```
Comment (carissa-tang) [inline frontend/src/components/common/Table.tsx]:
```suggestion
                backgroundColor={rowIndex % 2 === 0 ? "blue.50" : "grey.50"}
```

File: frontend/src/components/common/Table.tsx [hunk 20]
```diff
@@ -20,39 +21,59 @@ interface TableProps {
...
-            >
-              {row.values.map((value, cellIndex) => (
-                <Td key={value} fontWeight={cellIndex === 0 ? "bold" : ""}>
-                  {value}
-                </Td>
+    <>
+      <TableContainer
+        padding="0.5em"
+        border="1px solid"
+        borderColor="#E8EDF1"
```
Comment (carissa-tang) [inline frontend/src/components/common/Table.tsx]:
```suggestion
        borderColor="blue.50"
```

File: frontend/src/components/common/Table.tsx [hunk 21]
```diff
@@ -20,39 +21,59 @@ interface TableProps {
...
-    </TableContainer>
+          </Thead>
+          <Tbody>
+            {itemsToShow.map((row, rowIndex) => (
+              <Tr
+                key={rowIndex}
+                backgroundColor={rowIndex % 2 === 0 ? "#E8EDF1" : "#FFFFFF"}
+              >
+                {row.values.map((value, cellIndex) => (
+                  <Td key={value} fontWeight={cellIndex === 0 ? "500" : ""}>
```
Comment (carissa-tang) [inline frontend/src/components/common/Table.tsx]:
```suggestion
                  <Td key={value} fontWeight={cellIndex === 0 ? "bold" : ""}>
```
try not to change existing code!

File: frontend/src/components/common/Table.tsx [hunk 22]
```diff
@@ -20,39 +21,59 @@ interface TableProps {
...
+              </Tr>
+            ))}
+          </Tbody>
+        </T>
+      </TableContainer>
+      <Pagination
+        pagesCount={totalPages}
+        currentPage={currentPage}
+        onPageChange={setCurrentPage}
+        itemsToShow={10}
```
Comment (carissa-tang) [inline frontend/src/components/common/Table.tsx]:
```suggestion
```

File: frontend/src/components/common/Pagination.tsx [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+  currentPage: number;
+  itemsToShow: number;
+  onPageChange: (page: number) => void;
+}
+
+const Pagination = ({
+  pagesCount,
+  currentPage,
+  onPageChange,
+  itemsToShow,
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]: looks like this isn't used
```suggestion
```

File: frontend/src/components/common/Pagination.tsx [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+  PaginationSeparator,
+} from "@ajna/pagination";
+
+const outerLimit = 1;
+const innerLimit = 1;
+
+interface PaginationProps {
+  pagesCount: number;
+  currentPage: number;
+  itemsToShow: number;
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]:
```suggestion
```

File: frontend/src/components/common/Pagination.tsx [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+                return (
+                  <PaginationSeparator
+                    size="PaginationStyle"
+                    key={`pagination_ellipsis_${index}`}
+                    sx={{
+                      bg: "blue.50",
+                      color: "grey.300",
+                      height: "2.25rem",
+                      width: "2.25rem",
+                      borderRadius: "50%",
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]:
```suggestion
                      opacity: "1 !important",
                      bg: "blue.50",
                      color: "grey.300",
```
for the styling that is already defined in PaginationStyle, you don't need to redefine it (unless it's not applying like for the colours)

also! the reason the colour was lighter is because the button is disabled and applies `opacity: 0.4` - so we can override it!

File: frontend/src/components/common/Pagination.tsx [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+
+  return (
+    <Box
+      bottom="0"
+      left="0"
+      right="0"
+      display="flex"
+      justifyContent="center"
+      style={{ height: "75px" }}
+    >
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]: i think you can remove this box
```suggestion
```

File: frontend/src/components/common/Pagination.tsx [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+          </PaginationPageGroup>
+
+          {currentPage !== pagesCount && (
+            <PaginationNext size="paginationNext" variant="paginationNext">
+              Next &gt;
+            </PaginationNext>
+          )}
+        </PaginationContainer>
+      </P>
+    </Box>
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]:
```suggestion
```

File: frontend/src/components/user-management/AdminUserTable.tsx [hunk 28]
```diff
@@ -1,6 +1,7 @@
 import React from "react";
 import { AdminUser } from "../../types/UserTypes";
 import { TableRow, Table } from "../common/Table";
+
```
Comment (carissa-tang) [inline frontend/src/components/user-management/AdminUserTable.tsx]:
```suggestion
```

File: frontend/src/themes/components/button.ts [hunk 29]
```diff
@@ -27,8 +27,56 @@ const Button = {
...
+    },
+    paginationNext: {
```
Comment (carissa-tang) [inline frontend/src/themes/components/button.ts]: you can combine `paginationNext` and `paginationPrevious` into one size

File: frontend/src/components/common/Pagination.tsx [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+            if (
+              index === 0 ||
+              index === pages.length - 1 ||
+              (index >= currentPage - innerLimit - 1 &&
+                index <= currentPage + innerLimit - 1)
+            ) {
+              return (
+                <PaginationPage
+                  size="paginationStyle"
+                  variant="PaginationStyle"
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]:
```suggestion
                  variant="paginationStyle"
```

File: frontend/src/components/common/Pagination.tsx [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+                  key={`pagination_page_${page}`}
+                  page={page}
+                  _current={{
+                    bg: "blue.100",
+                    fontFamily: "DM Sans",
+                    borderRadius: "50%",
```
Comment (carissa-tang) [inline frontend/src/components/common/Pagination.tsx]:
```suggestion
```

**Review by carissa-tang — APPROVED** [review]
LGTM !!
